sql/database.go

```diff
@@ -292,16 +330,194 @@ func Open(uri string, opts ...Opt) (*sqliteDatabase, error) {
 	return db, nil
 }
 
+func ensureDBSchemaUpToDate(logger *zap.Logger, db *sqliteDatabase, config *conf) (*sqliteDatabase, error) {
+	before, after, err := config.schema.CheckDBVersion(logger, db)
+	switch {
+	case err != nil:
+		return db, err
+	case before == after:
+		return db, nil
+	case before > after:
+		// TODO: this should be logged by the caller
+		logger.Error("database version is newer than expected - downgrade is not supported",
+			zap.Int("current version", before),
+			zap.Int("target version", after),
+		)
+		return db, fmt.Errorf("%w: %d > %d", ErrTooNew, before, after)
+	case !config.enableMigrations:
+		// TODO: this should be logged by the caller
+		logger.Error("database version is too old",
+			zap.Int("current version", before),
+			zap.Int("target version", after),
+		)
+		return db, fmt.Errorf("%w: %d < %d", ErrOldSchema, before, after)
+	case config.temp:
+		// Temporary database, do migrations without transactions
+		// and sync afterwards
+		return db, config.schema.MigrateTempDB(logger, db, before)
+	case config.vacuumState != 0 &&
+		before <= config.vacuumState &&
+		strings.HasPrefix(config.uri, "file:"):
+		logger.Info("running migrations",
+			zap.Int("current version", before),
+			zap.Int("target version", after),
+		)
+		return db.copyMigrateDB(config)
+	default:
```

Comment:
I would argue this could be outside the `switch` case. Happy path shouldn't be nested ideally

Reply:
Happy path here is `before == after`, and I'm not sure extracting it from the `switch` will make things much cleaner. Other cases are "less happy" :)

Reply:
Since every `case` returns at the end you can just omit `default` and add the code after the `switch` block 🙂